### ISSUE TITLE
scipy.interpolate.griddata using method='linear' produces NaNs

### ISSUE THREAD

**Issue by ivanovmg** [scipy/scipy#10347]
My issue is about NaNs produced using scipy.interpolate.griddata.

Consider that the values from `values.csv` are defined for each point from `points.csv`.
We would like to get interpolated values for points from `xi.csv`.
In the following code snippet I check whether the interpolated values contain any NaNs.

### Reproducing code example:

```
from scipy.interpolate import griddata
import numpy as np
import pandas as pd

points = pd.read_csv('points.csv').values
values = pd.read_csv('values.csv').values
xi = pd.read_csv('xi.csv').values

interpolated = griddata(points, values, xi, method='linear', rescale=True)

idx_nan = np.where(np.isnan(interpolated))[0]
if idx_nan.any():
    print('NaNs are present')
    print(xi[idx_nan])
else:
    print('Success')

```

I used two sets of inputs for points and values: `with_duplicates` and `no_duplicates`.
In the former ones there are points with the duplicate coordinates, while in the other one they are eliminated.

#### With Duplicates

```
NaNs are present
[[ 2.   0.   0.  30. ]
 [ 2.   0.   0.  50. ]
 [ 2.   0.   0.  70. ]
 [ 2.5  0.   0.  30. ]
 [ 2.5  0.   0.  50. ]
 [ 2.5  0.   0.  70. ]
 [ 3.   0.   0.  30. ]
 [ 3.   0.   0.  50. ]
 [ 3.   0.   0.  70. ]
 [ 3.5  0.   0.  30. ]
 [ 3.5  0.   0.  50. ]
 [ 3.5  0.   0.  70. ]
 [ 4.   0.   0.  30. ]
 [ 4.   0.   0.  50. ]
 [ 4.   0.   0.  70. ]
 [ 4.5  0.   0.  30. ]
 [ 4.5  0.   0.  50. ]
 [ 4.5  0.   0.  70. ]
 [ 5.   0.   0.  30. ]
 [ 5.   0.   0.  50. ]
 [ 5.   0.   0.  70. ]]

```

I wonder why the NaNs were generated not for all points.
Note, that there were no NaNs for `[1 0 0 30]` point, for example.

The points, where we have NaNs as a results on interpolation are actually present in `points.csv`, for example `[5, 0, 0, 30]` is there.
Thus, it is surprising why interpolation led to NaNs even if there is a perfect match in the `points` array.

#### No Duplicates

```
Success
```

I removed duplicate points and re-ran the code snippet. Now it was a success.

It is not clear, though, why there would be such a behavior with `with_duplicates`. If the point from `xi` matches point from `points` array, then it seems that the algorithm should just pick the corresponding value.

### Scipy/Numpy/Python version information:

```
import sys, scipy, numpy; print(scipy.__version__, numpy.__version__, sys.version_info)
1.2.1 1.16.3 sys.version_info(major=3, minor=7, micro=3, releaselevel='final', serial=0)
```

## Files with data
[xi.zip](https://github.com/scipy/scipy/files/3321221/xi.zip)
[with_duplicates.zip](https://github.com/scipy/scipy/files/3321213/with_duplicates.zip)
[no_duplicates.zip](https://github.com/scipy/scipy/files/3321225/no_duplicates.zip)

**Comment by eslambakr** [comment]
Is there any updates regarding this issue?

**Comment by ev-br** [comment]
TL;DR: these nans come from QHull's triangulating the data; the `griddata` routine treats these nan-producing points as being outside and refuses to extrapolate.

Running locally the snippet from the OP with data from `with_duplicates.zip` and `xi.zip` above, produces two nans at indices 18 and 19:

```
In [1]: path = './'

In [2]: from scipy.interpolate import griddata
   ...: import numpy as np
   ...: import pandas as pd
   ...:
   ...: points = pd.read_csv(path + 'points.csv').values
   ...: values = pd.read_csv(path + 'values.csv').values

In [3]: xi = pd.read_csv(path + 'xi.csv').values

In [4]: interpolated = griddata(points, values, xi, method='linear', rescale=True)

In [5]: interpolated.T
Out[5]: 
array([[ 1.,  1.,  1.,  1.,  1.,  1.,  1.,  1.,  1.,  1.,  1.,  1.,  1.,
         1.,  1.,  1.,  1.,  1., nan, nan,  1.,  1.,  1.,  1.,  1.,  1.,
         1.]])
```

Changing the `fill_value` from the default nan to some other value changes precisely these elements:

```
In [9]: griddata(points, values, xi, method='linear', rescale=True, fill_value=42).T
Out[9]: 
array([[ 1.,  1.,  1.,  1.,  1.,  1.,  1.,  1.,  1.,  1.,  1.,  1.,  1.,
         1.,  1.,  1.,  1.,  1., 42., 42.,  1.,  1.,  1.,  1.,  1.,  1.,
         1.]])
```

To verify, using this patch

```diff
diff --git a/scipy/interpolate/interpnd.pyx b/scipy/interpolate/interpnd.pyx
index e323997500..0abefd54ca 100644
--- a/scipy/interpolate/interpnd.pyx
+++ b/scipy/interpolate/interpnd.pyx
@@ -324,6 +324,10 @@ class LinearNDInterpolator(NDInterpolatorBase):
 
                 if isimplex == -1:
                     # don't extrapolate
+
+                    with gil:
+                        print("isimplex=-1 at i=", i)
+
                     for k in range(nvalues):
                         out[i,k] = fill_value
                     continue
```
and reruning the snippets above, prints 

```
In [9]: griddata(points, values, xi, method='linear', rescale=True, fill_value=42).T
isimplex=-1 at i= 18
isimplex=-1 at i= 19
Out[9]: 
array([[ 1.,  1.,  1.,  1.,  1.,  1.,  1.,  1.,  1.,  1.,  1.,  1.,  1.,
         1.,  1.,  1.,  1.,  1., 42., 42.,  1.,  1.,  1.,  1.,  1.,  1.,
         1.]])
```

So, all in all, QHull gets confused by these duplicate points (as a guess, the corresponding simplices are degenerate) and griddata bails out with the `fill_value`. There is not much else scipy.interpolate can do really. 
issue
> It is not clear, though, why there would be such a behavior with with_duplicates. If the point from xi matches point from points array, then it seems that the algorithm should just pick the corresponding value.

This expectation is similar to https://github.com/scipy/scipy/issues/11381, and the answer is the same: no, data needs user preprocessing. Here it is actually even more clear-cut than in gh-11381. Imagine that the `values` are different for duplicate `points`. What should `griddata` do in that case? 

Conclusing, the answer is to filter out duplicates on the user side. There is nothing to track on the scipy.interpolate side, so I'm closing the issue. Feel free to keep discussing  however!

Thanks @ivanovmg , all.
